packages/storage-driver-cloudinary/src/index.ts

```diff
@@ -0,0 +1,387 @@
+import type { Driver, Range } from '@directus/storage';
+import { normalizePath } from '@directus/utils';
+import { Blob, Buffer } from 'node:buffer';
+import { createHash } from 'node:crypto';
+import { extname, join, parse } from 'node:path';
+import { Readable } from 'node:stream';
+import PQueue from 'p-queue';
+import type { RequestInit } from 'undici';
+import { fetch, FormData } from 'undici';
+import { IMAGE_EXTENSIONS, VIDEO_EXTENSIONS } from './constants.js';
+
+export type DriverCloudinaryConfig = {
+	root?: string;
+	cloudName: string;
+	apiKey: string;
+	apiSecret: string;
+	accessMode: 'public' | 'authenticated';
+};
+
+export class DriverCloudinary implements Driver {
+	private root: string;
+	private apiKey: string;
+	private apiSecret: string;
+	private cloudName: string;
+	private accessMode: 'public' | 'authenticated';
+
+	constructor(config: DriverCloudinaryConfig) {
+		this.root = config.root ? normalizePath(config.root, { removeLeading: true }) : '';
+		this.apiKey = config.apiKey;
+		this.apiSecret = config.apiSecret;
+		this.cloudName = config.cloudName;
+		this.accessMode = config.accessMode;
+	}
+
+	private fullPath(filepath: string) {
+		return normalizePath(join(this.root, filepath), { removeLeading: true });
+	}
+
+	private toFormUrlEncoded(obj: Record<string, string>, options?: { sort: boolean }) {
+		let entries = Object.entries(obj);
+
+		if (options?.sort) {
+			entries = entries.sort(([keyA], [keyB]) => keyA.localeCompare(keyB));
+		}
+
+		return new URLSearchParams(entries).toString();
+	}
+
+	/**
+	 * Generate the Cloudinary SHA1 signature for the given payload
+	 * @see https://cloudinary.com/documentation/signatures
+	 */
+	private getFullSignature(payload: Record<string, string>) {
+		const denylist = ['file', 'cloud_name', 'resource_type', 'api_key'];
+
+		const signaturePayload = Object.fromEntries(
+			Object.entries(payload).filter(([key]) => denylist.includes(key) === false)
+		);
+
+		const signaturePayloadString = this.toFormUrlEncoded(signaturePayload, { sort: true });
+
+		return createHash('sha1')
+			.update(signaturePayloadString + this.apiSecret)
+			.digest('hex');
+	}
+
+	/**
+	 * Creates inline URL signature for use with the image reading API
+	 * @see https://cloudinary.com/documentation/advanced_url_delivery_options#generating_delivery_url_signatures
+	 */
+	private getParameterSignature(filepath: string) {
+		return `s--${createHash('sha1')
+			.update(filepath + this.apiSecret)
+			.digest('base64url')
+			.substring(0, 8)}--`;
+	}
```

Comment:
This is also more of a possible edge case concern, but do we want to make it possible/customizable to use `sha256` as well? It seems like it is technically possible for someone to limit their account to only allow SHA-256 (refer quote below), so I presume said users will face issues when the driver is only using sha1 hash. However I'm not sure how common that is among actual Cloudinary users.

> Note: By default, Cloudinary supports both SHA-1 and SHA-256 digests for validation, and you can use either. The SDK methods use the SHA-1 algorithm by default, but you can use the SHA-256 algorithm instead by setting the `signature_algorithm` SDK [configuration parameter](https://cloudinary.com/documentation/cloudinary_sdks#configuration_parameters) to `sha256`. If you want to limit your account to allow only the SHA-256 digest for all your validations, [submit a request](https://support.cloudinary.com/hc/en-us/requests/new).

Link to above quote in an info block: https://cloudinary.com/documentation/upload_images#manual_signature_generation

Reply:
> However I'm not sure how common that is among actual Cloudinary users.

I think I feel the same way about this as the custom CDN url functionality. I don't know if this is common enough to warrant adding right now. This PR is already absolutely massive, so I'd rather launch this a little simpler, and circle back to that if the need ever comes up 🙂

Reply:
You know what, nevermind! Turns out cloudinary supports both sha-1 and sha-256 by default, and you can optionally limit it to only sha-256 if you are enterprise. That means we should be able to use sha-256 at all times and make it work for everybody at all times 🙂 

> By default, Cloudinary supports both SHA-1 and SHA-256 digests for validation, and you can use either

https://cloudinary.com/documentation/signatures

Reply:
> That means we should be able to use sha-256 at all times and make it work for everybody at all times 🙂

That's actually way better and straightforward 🤯